fix #51135 make CFS quota period configurable

szuecs opened the pull request:
**What this PR does / why we need it**:

This PR makes it possible for users to change CFS quota period from the default 100ms to some other value between 1µs and 1s.
#51135 shows that multiple production users have serious issues running reasonable workloads in kubernetes. The latency added by the 100ms CFS quota period is adding way too much time.

**Which issue(s) this PR fixes**:
Fixes #51135 

**Special notes for your reviewer**:
- 5ms is used by user experience https://github.com/kubernetes/kubernetes/issues/51135#issuecomment-384908627
- Latency added caused by CFS 100ms is shown at https://github.com/kubernetes/kubernetes/issues/51135#issuecomment-373454012
- explanation why we should not disable limits https://github.com/kubernetes/kubernetes/issues/51135#issuecomment-385346661
- agreement found at kubecon EU 2018: https://github.com/kubernetes/kubernetes/issues/51135#issuecomment-386623964

**Release note**:
```release-note
Adds a kubelet parameter and config option to change CFS quota period from the default 100ms to some other value between 1µs and 1s. This was done to improve response latencies for workloads running in clusters with guaranteed and burstable QoS classes.  
```

szuecs committed:
fix #51135 make CFS quota period configurable, adds a cli flag and config option to kubelet to be able to set cpu.cfs_period and defaults to 100ms as before.
It requires to enable feature gate CustomCPUCFSQuotaPeriod.

Signed-off-by: Sandor Szücs <sandor.szuecs@zalando.de>

liggitt commented:
@kubernetes/sig-node-pr-reviews  @derekwaynecarr

dims commented:
@derekwaynecarr The `"cpu-cfs-quota"` command line flag in kubelet right? So we would have a `"cpu-cfs-quota-period"` to go along with it?

derekwaynecarr commented:
Yes

On Fri, May 4, 2018 at 5:37 PM Davanum Srinivas <notifications@github.com>
wrote:

> @derekwaynecarr <https://github.com/derekwaynecarr> The "cpu-cfs-quota"
> command line flag in kubelet right? So we would have a
> "cpu-cfs-quota-period" to go along with it?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/63437#issuecomment-386741730>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF8dbPLiCmJFAt46M0b_94jS34wGFFxYks5tvMohgaJpZM4Ty3li>
> .
>

dims commented:
@szuecs looking forward to an update in this PR with the suggestion above

derekwaynecarr commented:
see my comment here for why a flag is preferred.  Not all nodes are tuned for latency.

https://github.com/kubernetes/kubernetes/issues/51135#issuecomment-386767390

szuecs commented:
Thanks for all your comments.

szuecs commented:
@derekwaynecarr I added the cli flag and config option. I hope that I added it everywhere, where it is needed.

I tested, that:
- tests run and report ok % go test ./pkg/kubelet/cm -v 
- I can locally build kubelet: % make WHAT=cmd/kubelet

Let me know if I have to change anything.

vishh commented:
/hold
I'd like for us to reach a consensus on https://github.com/kubernetes/kubernetes/issues/51135#issuecomment-387139466 prior to merging this patch.

szuecs commented:
/retest

dims commented:
@derekwaynecarr @vishh Do we count this as a new feature? if so, Does this feature need an alpha feature gate?

dims commented:
@derekwaynecarr @vishh Also, do we leave the current default as-is? (when nothing is specified in the command line)

dims commented:
@szuecs 

- if run "hack/update-bazel.sh" that should take care of the verify job failure. (and bazel-test job failure too)
- the fix for e2e failure should have already gone in, but you may have to rebase to master

So, please squash the commits (for easier review), rebase to master and we should get all green (cross my fingers)

thanks,
Dims

szuecs commented:
@dims The update-bazel.sh did no change anything.
 
Without run-in-gopath.sh I get panics all the time, as far as I understood the developer docs I should run it with run-in-gopath.sh. 
I fetched master and merged into the branch, such that is up to date.
I also pulled the godeps with hack/run-in-gopath.sh hack/godep-restore.sh, but it does not change the update-bazel run:

```
[sszuecs@sandor-lab:~/go/src/k8s.io/kubernetes]% hack/run-in-gopath.sh hack/update-bazel.sh
[sszuecs@sandor-lab:~/go/src/k8s.io/kubernetes]% git st
On branch fix/51135-set-saneer-default-cpu.cfs_period
nothing to commit, working directory clean
```

Any hints?

dims commented:
/retest

dims commented:
/assign @derekwaynecarr 
/assign @vishh 
/assign @liggitt 

Questions:
- Do we need a alpha feature gate?
- Anything specific we need to do for an upgrade scenario?

dims reviewed (approved):
Looks good to me so far. thanks for the rebase + squash.

timothysc reviewed (approved):
/approve

derekwaynecarr commented:
I am in favor of this change, but would like to discuss this topic in tomorrow's sig-node meeting in order to reach consensus on the path forward.

/cc @dchen1107

derekwaynecarr commented:
@dims - this should not demand a feature gate.

derekwaynecarr commented on cmd/kubelet/app/options/options.go:
```diff
@@ -535,6 +535,7 @@ func AddKubeletConfigFlags(mainfs *pflag.FlagSet, c *kubeletconfig.KubeletConfig
 
 	fs.StringVar(&c.ResolverConfig, "resolv-conf", c.ResolverConfig, "Resolver configuration file used as the basis for the container DNS resolution configuration.")
 	fs.BoolVar(&c.CPUCFSQuota, "cpu-cfs-quota", c.CPUCFSQuota, "Enable CPU CFS quota enforcement for containers that specify CPU limits")
+	fs.DurationVar(&c.CPUCFSQuotaPeriod.Duration, "cpu-cfs-quota-period", c.CPUCFSQuotaPeriod.Duration, "Sets CPU CFS quota period value, cpu.cfs_period_us, defaults to 100ms")
```
please add that this is the kernel default.

liggitt replied:
Default is printed automatically, no need to double include in help text

szuecs replied:
Changed as commented

dims commented:
/retest

szuecs commented:
@derekwaynecarr I think you have to add your approval, too as far as I understand. Let me know if there is something missing or you just need time to review. Thanks

derekwaynecarr commented:
@szuecs - i will review the code in depth later today or tomorrow.

derekwaynecarr commented:
Hmm, the plot thickens on this PR.

It appears this PR works with the `cgroupfs` cgroup driver and not the `systemd` cgroup driver.

The issue is that when using systemd, systemd only exposes in its dbus API `CPUQuotaPerSecUSec`.

see in runc:
https://github.com/opencontainers/runc/blob/master/libcontainer/cgroups/systemd/apply_systemd.go#L314

and in systemd the period is fixed via constant:
https://github.com/systemd/systemd/blob/81183d9b9934bc508820d7bedb82695c5f74ed27/src/core/cgroup.c#L517

i will reach out to systemd community to see if i am missing something, but we should think about how to proceed with this pr.  if we ever wanted to support a per pod period tuning option, i would want to have that knob settable via systemd, or know its safe to set it behind systemd before merging this since the systemd cgroup driver is GA.

in the interim, we may want to mark this flag as experimental, and disable the ability to update the value when using systemd cgroup driver.

derekwaynecarr commented:
I have opened the following issue with systemd: https://github.com/systemd/systemd/issues/9081

hjacobs commented:
@derekwaynecarr @BenPope thanks! @szuecs could you take a look at https://github.com/szuecs/kubernetes/pull/1?

szuecs commented:
@BenPope I merged your PR into this PR and squashed the commit. Thanks!

szuecs commented:
/retest

szuecs commented:
/retest

dims commented:
@szuecs looks like there are some legit failures in the ci jobs

szuecs commented:
@dims thanks I will check it later today or at least this week. It's not always clear to me when I just have to retest and when it's a real issue.

szuecs commented:
/test pull-kubernetes-bazel-build

derekwaynecarr commented:
it looks like this is a real issue, we can plan for this change in 1.12

szuecs commented:
@derekwaynecarr fixed

dims commented:
/hold cancel

/milestone v1.12 

( per last comment from @derekwaynecarr )

szuecs commented:
kind/feature

dims commented:
/kind feature

dims commented:
/assign @derekwaynecarr

let's please get this in early in the cycle now that the master is open

feiskyer commented:
/priority important-soon

feiskyer commented:
/status approved-for-milestone

feiskyer commented on cmd/kubelet/app/options/options.go:
```diff
@@ -535,6 +535,7 @@ func AddKubeletConfigFlags(mainfs *pflag.FlagSet, c *kubeletconfig.KubeletConfig
 
 	fs.StringVar(&c.ResolverConfig, "resolv-conf", c.ResolverConfig, "Resolver configuration file used as the basis for the container DNS resolution configuration.")
 	fs.BoolVar(&c.CPUCFSQuota, "cpu-cfs-quota", c.CPUCFSQuota, "Enable CPU CFS quota enforcement for containers that specify CPU limits")
+	fs.DurationVar(&c.CPUCFSQuotaPeriod.Duration, "cpu-cfs-quota-period", c.CPUCFSQuotaPeriod.Duration, "Sets CPU CFS quota period value, cpu.cfs_period_us, defaults to Linux Kernel default")
```
`defaults to Linux Kernel default (100ms)` for clear?

szuecs replied:
See the former comment from ligitt. Cobra will render the default automatically.

dixudx replied:
We should stop adding new flags to kubelet. Add directly to `KubeletConfiguration` instead?

derekwaynecarr replied:
i would prefer to not have the flag and just use via `KubeletConfiguration`; we should note that dynamically changing the field value will not impact currently running pods (only new pods)

feiskyer reviewed (commented):
LGTM

szuecs commented:
/test pull-kubernetes-local-e2e-containerized

hjacobs commented:
/lgtm

szuecs commented:
@derekwaynecarr @vishh anything that stops this PR from getting a lgtm?
I feel a bit sick of rebasing and fixing merge conflicts and tests, because of refactorings in other places.

derekwaynecarr commented:
Can we put this behind a feature gate flag?  The systemd community has prepared a patch to add support for variable quota , but since this option is not yet supported across all cgroup drivers, I would prefer to put it behind a feature gate flag in the interim.

szuecs commented:
@derekwaynecarr can you roughly explain where this is done?
I guess it's a  value in a kubelet flag, right?

dims commented:
@szuecs you can follow the same pattern as [`SupportPodPidsLimit`](https://cs.k8s.io/?q=SupportPodPidsLimit&i=nope&files=&repos=kubernetes)

szuecs commented:
I added feature gate CustomCPUCFSQuotaPeriod to enable the feature. 
I will check tomorrow, if the tests are passing.

szuecs commented:
/retest

szuecs commented:
/retest

dims commented:
/retest

szuecs commented:
/test pull-kubernetes-e2e-gke

szuecs commented:
/test pull-kubernetes-e2e-gce

szuecs commented:
/test pull-kubernetes-e2e-gce-100-performance

szuecs commented:
/test pull-kubernetes-kubemark-e2e-gce-big

szuecs commented:
/test pull-kubernetes-e2e-gke
/test pull-kubernetes-e2e-gce
/test pull-kubernetes-e2e-gce-100-performance

szuecs commented:
/test pull-kubernetes-e2e-gce

xiangpengzhao commented:
/retest

szuecs commented:
/retest

szuecs commented:
/retest

szuecs commented:
/retest

szuecs commented:
/retest

dims commented on pkg/kubelet/kuberuntime/helpers_linux.go:
```diff
@@ -54,11 +52,8 @@ func milliCPUToQuota(milliCPU int64) (quota int64, period int64) {
 		return
 	}
```
@szuecs Don't we have to use old behavior here when the alpha feature flag is not set?

```
	if !utilfeature.DefaultFeatureGate.Enabled(kubefeatures.CPUCFSQuotaPeriod) {
		period = quotaPeriod
	}
```

szuecs replied:
I added additional unit tests to make sure we have the right values if feature gate is enabled or disabled

dims commented on pkg/kubelet/kuberuntime/helpers_linux.go:
```diff
@@ -24,8 +24,6 @@ const (
 	sharesPerCPU  = 1024
 	milliCPUToCPU = 1000
 
-	// 100000 is equivalent to 100ms
```
We need this below

dims commented:
/test pull-kubernetes-e2e-gke
/test pull-kubernetes-cross
/test pull-kubernetes-e2e-kubeadm-gce
/test pull-kubernetes-local-e2e

Run the skipped tests just to be sure!

szuecs commented:
@derekwaynecarr PR is ready for review, feature gates set up, more unit tests added, all e2e tests green.

dims commented:
/lgtm

dims commented:
/assign @yujuhong 
/assign @Random-Liu

szuecs commented:
/test pull-kubernetes-integration

dims commented:
/retest

szuecs commented:
/retest

timothysc reviewed (approved):
/approve - from my side

derekwaynecarr commented on pkg/kubelet/apis/kubeletconfig/v1beta1/types.go:
```diff
@@ -453,6 +453,10 @@ type KubeletConfiguration struct {
+	// cpuCFSQuotaPeriod is the CPU CFS quota period value, cpu.cfs_period_us
+	// Default: "100ms"
+	// +optional
+	CPUCFSQuotaPeriod *metav1.Duration `json:"cpuCFSQuotaPeriod,omitempty"`
```
can you add a note about the impact of changing this field when using Dynamic Kubelet Config?

szuecs replied:
done

derekwaynecarr commented on pkg/kubelet/apis/kubeletconfig/v1beta1/defaults.go:
```diff
@@ -151,6 +151,9 @@ func SetDefaults_KubeletConfiguration(obj *KubeletConfiguration) {
 	if obj.CPUCFSQuota == nil {
 		obj.CPUCFSQuota = utilpointer.BoolPtr(true)
 	}
+	if obj.CPUCFSQuotaPeriod == nil && obj.FeatureGates["CustomCPUCFSQuotaPeriod"] {
```
reference the feature gate name from features package directly.

derekwaynecarr commented on pkg/kubelet/apis/kubeletconfig/validation/validation.go:
```diff
@@ -51,6 +52,9 @@ func ValidateKubeletConfiguration(kc *kubeletconfig.KubeletConfiguration) error
 	if kc.HealthzPort != 0 && utilvalidation.IsValidPortNum(int(kc.HealthzPort)) != nil {
 		allErrors = append(allErrors, fmt.Errorf("invalid configuration: HealthzPort (--healthz-port) %v must be between 1 and 65535, inclusive", kc.HealthzPort))
 	}
+	if localFeatureGate.Enabled("CustomCPUCFSQuotaPeriod") && utilvalidation.IsInRange(int(kc.CPUCFSQuotaPeriod.Duration), int(1*time.Microsecond), int(time.Second)) != nil {
```
same here reference `features.` feature name directly.

derekwaynecarr commented on pkg/kubelet/cm/helpers_linux_test.go:
```diff
@@ -49,21 +53,30 @@ func getResourceRequirements(requests, limits v1.ResourceList) v1.ResourceRequir
+	tunedQuotaPeriod := uint64(5 * time.Millisecond / time.Microsecond)
+	tunedQuota := int64(1 * time.Millisecond / time.Microsecond)
+
+	utilfeaturetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, pkgfeatures.CPUCFSQuotaPeriod, true)
```
doesn't this stop us from testing the function when the feature gate is off?

shouldn't you have a seperate Test function that runs the test with it off, and another with it on?

szuecs replied:
I duplicated the test and have one enabled and one disabled.
One problem for the test is that the ResourceConfigForPod() gets the period as paramter so the test will have expected the tunedQuotaPeriod because of that. The period parameter will not be passed into it, but this is outside of the cm package. 
I am not sure if this is nice to read in the test, maybe you have a better suggestion? 
I could add the feature gate in cm, such that the parameter will be ignored, which I think makes even less sense.

derekwaynecarr commented on pkg/kubelet/kuberuntime/helpers_linux.go:
```diff
@@ -44,21 +49,23 @@ func milliCPUToShares(milliCPU int64) int64 {
-
-	// we set the period to 100ms by default
-	period = quotaPeriod
+	println("before feature gate period:", period)
```
remove this debugging

derekwaynecarr reviewed (changes requested):
i have a number of code nits, but how do you feel about writing a node e2e test as well that validates the function?  the test can run if not using systemd cgroup driver, and work similar to pods_container_manager_test ?

szuecs commented:
@derekwaynecarr I changed as commented, but see https://github.com/kubernetes/kubernetes/pull/63437#discussion_r212381282, which could be odd. 
I did not add the e2e test, yet.

derekwaynecarr commented:
thank you for the the updated test case, i would prefer an e2e test, but i wont hold this pr on that.

if we ever do graduate the feature to beta, it should definitely have tests.

/approve
/lgtm

derekwaynecarr commented:
/retest

szuecs commented:
/retest

szuecs commented:
/retest

dims commented:
reapplying lgtm from @derekwaynecarr (lost in rebase)

/lgtm

hjacobs commented:
@derekwaynecarr @dims @timothysc @szuecs thanks!

dashpole commented:
This has broken the release blocking alphafeatures test suite: https://k8s-testgrid.appspot.com/sig-release-1.12-blocking#gce-cos-1.12-alphafeatures

The kubelet is crashlooping:
server.go:207] invalid configuration: CPUCFSQuotaPeriod (--cpu-cfs-quota-period) {0s} must be between 1usec and 1sec, inclusive

Tracking bug for release-blocking failures: https://github.com/kubernetes/kubernetes/issues/68313

dashpole commented on pkg/kubelet/apis/config/v1beta1/defaults.go:
```diff
@@ -154,6 +155,9 @@ func SetDefaults_KubeletConfiguration(obj *KubeletConfiguration) {
 	if obj.CPUCFSQuota == nil {
 		obj.CPUCFSQuota = utilpointer.BoolPtr(true)
 	}
+	if obj.CPUCFSQuotaPeriod == nil && obj.FeatureGates[string(features.CPUCFSQuotaPeriod)] {
```
I don't know if this actually works...  We generally do not feature gate the defaulting of flags.  We generally only place the feature gate around code we don't want to run, rather than around configuration.

derekwaynecarr replied:
This does look like a problem.  The flag should always default, it just shouldn’t have been used.  Apologies for missing in review

guineveresaenger commented:
@dashpole @derekwaynecarr @dims please give this your immediate attention re: #68313.

dashpole commented:
Fix is out: https://github.com/kubernetes/kubernetes/pull/68386

warmchang commented:
ping @Pingan2017

bartoszhernas commented:
Our php container started seeing problems connected with CPU limits.

We started profiling and saw that `initSoapClient` method which usually takes 10ms, started taking up to 120s (no mistake, 2 minutes...). We do not use any container limits, we only use requests. After requesting for much more CPU than possibly needed (40%) the issue disappeared.

We are hosted on GKE, is there an option to disable cpu limits or reduce CFS period there? Does anyone has idea how to fix this problem on GKE?

paskal commented on pkg/kubelet/apis/config/types.go:
```diff
@@ -220,6 +220,8 @@ type KubeletConfiguration struct {
 	// cpuCFSQuota enables CPU CFS quota enforcement for containers that
 	// specify CPU limits
 	CPUCFSQuota bool
+	// CPUCFSQuotaPeriod sets the CPU CFS quota period value, cpu.cfs_period_us, defaults to 100ms
```
I'm super confused about this one and can't figure this one out by myself: `cpu.cfs_period_us` default limit called "100ms" but [actually](https://www.kernel.org/doc/html/latest/scheduler/sched-bwc.html#management) is microseconds:
> `cpu.cfs_period_us`: the length of a period (in microseconds)

In the change I'm commenting author clearly acted under assumptions that milliseconds are used in k8s and now it's all over the test cases and comments. But I wonder where that 1000x disperancy between k8s value and Linux value comes from? And does it exist, or k8s default is actually a microsecond, same as linux? Unless we multiple [that](https://github.com/kubernetes/kubernetes/blob/37761a3/pkg/kubelet/cm/helpers_linux.go#L47-L48) default value of `QuotaPeriod` by 1000 somewhere, converted to `time.Duration` it's 100 microseconds and not milliseconds.

cc @vishh @sjenning and especially @hjacobs: I hope you all have some context about it and can tell me why I'm wrong and k8s code comments are right.

szuecs replied:
yes cfs_period_us is in µs, that's why it has `_us` suffix
Maybe https://go.dev/play/p/65fG1OmLFYN helps to understand.
As far as I remember, I think the time.Duration is only used if you want to set it via kubelet flag/config.

paskal replied:
The comment in that MR says "ms"." In contrast, it better be "microseconds", I assume. I can send MR to change it as well as explicit millisecond mentions in the current code, whereas it should be microseconds. I'll ping you in it.

In this MR, you wrote milliseconds duration in multiple places in defaults and tests `CPUCFSQuotaPeriod: metav1.Duration{Duration: 100 * time.Millisecond},`, I wonder if that should be changed to microseconds to match the actual default value of 100 microseconds?

Also, the discussion which led to this MR has multiple people talking about changing the value from 100 milliseconds to 5/25 milliseconds for testing. Still, it doesn't make sense unless everyone were altering the source code in microseconds, thinking it's milliseconds. @hjacobs even did [a talk](https://youtu.be/eBChCFD9hfs?t=870) about it talking about the default value as if it was in milliseconds. Your code introducing milliseconds and not microseconds (so 1000x) contributed to my confusion, and I wonder if it's intentional or if you are mistaken, and it should be changed microseconds to change the real default.

The only way current code would make sense to me is if the value you feed to k8s (100 milliseconds) is divided by 1000 at some point in time and then matches the real Linux default of 100 microseconds, but I haven't found such a place yet, and I would doubt that's the case.

paskal replied:
@szuecs, the simplest form of the question I can form is the following.

Is the 100ms value of `CPUCFSQuota` result in the same k8s behaviour as the default Linux value of 100 microseconds, or is it 1000 times higher as its milliseconds versus microseconds in Linux?

paskal replied:
I was wrong, and the original code was correct. 100ms is indeed a default value in Linux, and I got confused by conversions in two places in code that had no comments about them. PR #112123 adds comments to the code added here.

paskal commented on pkg/kubelet/kuberuntime/helpers_linux.go:
```diff
@@ -44,21 +49,22 @@ func milliCPUToShares(milliCPU int64) int64 {
-	// we set the period to 100ms by default
-	period = quotaPeriod
+	if !utilfeature.DefaultFeatureGate.Enabled(kubefeatures.CPUCFSQuotaPeriod) {
+		period = quotaPeriod
```
@szuecs that's precise place which worries me: previously we used `quotaPeriod` as-is, which, translated to `time.Duration`, equals to 100 microseconds. And now we set it to `cpuCFSQuotaPeriod`, which according to code and comments is 100 millisecond default.

szuecs replied:
1. I don't see how this code can be worrying, because it shows the same assignment as before. If not the feature gate is set, set it to quotaPeriod as before. see https://github.com/kubernetes/kubernetes/pull/63437/files/588d2808b77d11f235b6eba5c21bcaa89a2f7804#r932625167
2. if you enable the featuregate you can set to the value you like

paskal replied:
"ms" mentioned across the code alongside the tests with milliseconds and not microseconds confused me. Now I see it's still microseconds, and comments should be altered, and maybe tests as well. I'll prepare the PR, thanks for the help!

szuecs replied:
Thanks @paskal , makes sense to update, if it's wrong :)

paskal replied:
I've created #111520 about it.

paskal replied:
I was wrong, and the original code was correct. 100ms is indeed a default value in Linux, and I got confused by conversions in two places in code that had no comments about them. PR #112123 adds comments to the code added here.

szuecs replied:
I think I read the kernel documentation at least 5 times when I did investigation and the code here.

szuecs commented on pkg/kubelet/kuberuntime/helpers_linux.go:
```diff
@@ -18,14 +18,19 @@ limitations under the License.
 	// 100000 is equivalent to 100ms
-	quotaPeriod    = 100 * minQuotaPeriod
+	quotaPeriod    = 100000
 	minQuotaPeriod = 1000
```
quotaPeriod calculation was inlined:
100 * minQuotaPeriod = 100 * 1000 = 100000
